File: src/api/passwords.py
"""
BiPro API - Passwort-Verwaltung

Zentrale Verwaltung bekannter Passwoerter fuer PDF- und ZIP-Entschluesselung.
Oeffentlicher Endpunkt fuer Desktop-Client + Admin-CRUD.
"""

from typing import Optional, Dict, List
import logging

from .client import APIClient, APIError

logger = logging.getLogger(__name__)


class PasswordsAPI:
    """
    API-Client fuer die Passwort-Verwaltung.
    
    Verwendung:
        passwords = PasswordsAPI(client)
        pdf_pws = passwords.get_passwords('pdf')
        zip_pws = passwords.get_passwords('zip')
    """
    
    def __init__(self, client: APIClient):
        self.client = client
    
    # ================================================================
    # Oeffentliche Endpunkte (JWT erforderlich, kein Admin)
    # ================================================================
    
    def get_passwords(self, password_type: str) -> List[str]:
        """
        Aktive Passwoerter nach Typ abrufen.
        
        Args:
            password_type: 'pdf' oder 'zip'
            
        Returns:
            Liste von Passwort-Strings
        """
        try:
            response = self.client.get(f'/passwords?type={password_type}')
            if response.get('success'):
                return response['data'].get('passwords', [])
        except APIError as e:
            logger.error(f"Fehler beim Laden der {password_type}-Passwoerter: {e}")
            raise
        return []
    
    # ================================================================
    # Admin-Endpunkte (Admin-Rechte erforderlich)
    # ================================================================
    
    def get_all_passwords(self, password_type: str = None) -> List[Dict]:
        """
        Alle Passwoerter auflisten (Admin).
        
        Args:
            password_type: Optional - 'pdf' oder 'zip' zum Filtern
            
        Returns:
            Liste von Passwort-Dicts mit id, password_type, password_value, 
            description, created_at, created_by, is_active
        """
        try:
            url = '/admin/passwords'
            if password_type:
                url += f'?type={password_type}'
            response = self.client.get(url)
            if response.get('success'):
                return response['data'].get('passwords', [])
        except APIError as e:
            logger.error(f"Fehler beim Laden der Passwoerter (Admin): {e}")
            raise
        return []
    
    def create_password(self, password_type: str, password_value: str,
                        description: str = '') -> Dict:
        """
        Neues Passwort anlegen (Admin).
        
        Args:
            password_type: 'pdf' oder 'zip'
            password_value: Das Passwort
            description: Optionale Beschreibung
            
        Returns:
            Das erstellte Passwort-Dict
        """
        data = {
            'password_type': password_type,
            'password_value': password_value,
            'description': description
        }
        try:
            response = self.client.post('/admin/passwords', json_data=data)
            if response.get('success'):
                return response['data'].get('password', {})
        except APIError as e:
            logger.error(f"Fehler beim Anlegen des Passworts: {e}")
            raise
        return {}
    
    def update_password(self, password_id: int, password_value: str = None,
                        description: str = None, is_active: bool = None) -> Dict:
        """
        Passwort bearbeiten (Admin).
        
        Args:
            password_id: ID des Passworts
            password_value: Neuer Passwort-Wert (optional)
            description: Neue Beschreibung (optional)
            is_active: Aktiv-Status (optional)
            
        Returns:
            Das aktualisierte Passwort-Dict
        """
        data = {}
        if password_value is not None:
            data['password_value'] = password_value
        if description is not None:
            data['description'] = description
        if is_active is not None:
            data['is_active'] = 1 if is_active else 0
        
        try:
            response = self.client.put(f'/admin/passwords/{password_id}', json_data=data)
            if response.get('success'):
                return response['data'].get('password', {})
        except APIError as e:
            logger.error(f"Fehler beim Aktualisieren des Passworts {password_id}: {e}")
            raise
        return {}
    
    def delete_password(self, password_id: int) -> bool:
        """
        Passwort deaktivieren / Soft-Delete (Admin).
        
        Args:
            password_id: ID des Passworts
            
        Returns:
            True bei Erfolg
        """
        try:
            response = self.client.delete(f'/admin/passwords/{password_id}')
            return response.get('success', False)
        except APIError as e:
            logger.error(f"Fehler beim Deaktivieren des Passworts {password_id}: {e}")
            raise

File: BiPro-Webspace Spiegelung Live/api/passwords.php
<?php
/**
 * BiPro API - Bekannte Passwoerter (PDF / ZIP)
 * 
 * Zentrale Verwaltung von Passwoertern fuer automatisches Entschluesseln
 * von PDF- und ZIP-Dateien beim Upload.
 * 
 * Oeffentliche Endpunkte (JWT erforderlich):
 * - GET /passwords?type=pdf       - Aktive Passwoerter nach Typ abrufen
 * - GET /passwords?type=zip       - Aktive ZIP-Passwoerter abrufen
 * 
 * Admin-Endpunkte (Admin-Rechte erforderlich):
 * - GET    /admin/passwords           - Alle Passwoerter auflisten
 * - GET    /admin/passwords?type=pdf  - Gefiltert nach Typ
 * - POST   /admin/passwords           - Neues Passwort anlegen
 * - PUT    /admin/passwords/{id}      - Passwort bearbeiten
 * - DELETE /admin/passwords/{id}      - Passwort deaktivieren (Soft-Delete)
 */

require_once __DIR__ . '/lib/db.php';
require_once __DIR__ . '/lib/jwt.php';
require_once __DIR__ . '/lib/response.php';
require_once __DIR__ . '/lib/permissions.php';
require_once __DIR__ . '/lib/activity_logger.php';
require_once __DIR__ . '/lib/crypto.php';

/**
 * Oeffentlicher Endpunkt: GET /passwords?type=pdf|zip
 * Gibt nur die Passwort-Werte zurueck (fuer Desktop-Client).
 */
function handlePasswordsPublicRequest(string $method): void {
    if ($method !== 'GET') {
        json_error('Methode nicht erlaubt', 405);
    }
    
    // JWT erforderlich (aber kein Admin)
    $payload = JWT::requireAuth();
    
    $type = $_GET['type'] ?? '';
    if (!in_array($type, ['pdf', 'zip'])) {
        json_error('Parameter "type" erforderlich (pdf oder zip)', 400);
    }
    
    $passwords = Database::query(
        'SELECT password_value FROM known_passwords WHERE password_type = ? AND is_active = 1 ORDER BY id ASC',
        [$type]
    );
    
    // SV-006 Fix: Passwoerter entschluesseln (verschluesselt gespeichert)
    $values = [];
    foreach ($passwords as $row) {
        try {
            $values[] = Crypto::decrypt($row['password_value']);
        } catch (Exception $e) {
            // Fallback: Klartext-Wert (fuer noch nicht migrierte Eintraege)
            $values[] = $row['password_value'];
        }
    }
    
    json_success(['passwords' => $values]);
}

/**
 * Admin-Endpunkt: /admin/passwords
 */
function handleAdminPasswordsRequest(?string $idOrAction, string $method): void {
    $payload = requireAdmin();
    
    switch ($method) {
        case 'GET':
            handleListPasswords();
            break;
            
        case 'POST':
            handleCreatePassword($payload);
            break;
            
        case 'PUT':
            if (empty($idOrAction) || !is_numeric($idOrAction)) {
                json_error('Passwort-ID erforderlich', 400);
            }
            handleUpdatePassword((int)$idOrAction, $payload);
            break;
            
        case 'DELETE':
            if (empty($idOrAction) || !is_numeric($idOrAction)) {
                json_error('Passwort-ID erforderlich', 400);
            }
            handleDeletePassword((int)$idOrAction, $payload);
            break;
            
        default:
            json_error('Methode nicht erlaubt', 405);
    }
}

/**
 * GET /admin/passwords[?type=pdf|zip]
 * Alle Passwoerter auflisten (mit optionalem Typ-Filter).
 */
function handleListPasswords(): void {
    $type = $_GET['type'] ?? '';
    
    if ($type && in_array($type, ['pdf', 'zip'])) {
        $passwords = Database::query(
            'SELECT id, password_type, password_value, description, created_at, created_by, is_active 
             FROM known_passwords 
             WHERE password_type = ?
             ORDER BY password_type ASC, id ASC',
            [$type]
        );
    } else {
        $passwords = Database::query(
            'SELECT id, password_type, password_value, description, created_at, created_by, is_active 
             FROM known_passwords 
             ORDER BY password_type ASC, id ASC'
        );
    }
    
    // SV-006 Fix: Passwoerter fuer Admin-Anzeige entschluesseln
    foreach ($passwords as &$pw) {
        try {
            $pw['password_value'] = Crypto::decrypt($pw['password_value']);
        } catch (Exception $e) {
            // Klartext-Wert (noch nicht migriert) - unveraendert lassen
        }
    }
    unset($pw);
    
    json_success(['passwords' => $passwords]);
}

/**
 * POST /admin/passwords
 * Body: { password_type: "pdf"|"zip", password_value: "...", description?: "..." }
 */
function handleCreatePassword(array $adminPayload): void {
    $data = get_json_body();
    require_fields($data, ['password_type', 'password_value']);
    
    $type = $data['password_type'];
    $value = trim($data['password_value']);
    $description = trim($data['description'] ?? '');
    
    // Validierung
    if (!in_array($type, ['pdf', 'zip'])) {
        json_error('Ungueltiger Typ. Erlaubt: pdf, zip', 400);
    }
    
    if (strlen($value) === 0) {
        json_error('Passwort darf nicht leer sein', 400);
    }
    
    if (strlen($value) > 255) {
        json_error('Passwort darf maximal 255 Zeichen lang sein', 400);
    }
    
    // SV-006 Fix: Duplikat-Pruefung muss alle vorhandenen Passwoerter entschluesseln und vergleichen
    $allOfType = Database::query(
        'SELECT id, password_value, is_active FROM known_passwords WHERE password_type = ?',
        [$type]
    );
    $existing = null;
    foreach ($allOfType as $row) {
        $storedValue = $row['password_value'];
        try {
            $storedValue = Crypto::decrypt($storedValue);
        } catch (Exception $e) {
            // Klartext-Wert (noch nicht migriert)
        }
        if ($storedValue === $value) {
            $existing = $row;
            break;
        }
    }
    
    if ($existing) {
        if (!$existing['is_active']) {
            // Deaktiviertes Passwort reaktivieren
            Database::execute('UPDATE known_passwords SET is_active = 1, description = ? WHERE id = ?',
                [$description, $existing['id']]);
            
            ActivityLogger::logAdmin($adminPayload, 'password_reactivated', 'known_password', (int)$existing['id'],
                "Passwort reaktiviert: {$type}",
                ['password_type' => $type, 'description' => $description]
            );
            
            $password = Database::queryOne('SELECT * FROM known_passwords WHERE id = ?', [$existing['id']]);
            json_success(['password' => $password], 'Passwort reaktiviert');
        }
        json_error('Dieses Passwort existiert bereits fuer diesen Typ', 409);
    }
    
    // SV-006 Fix: Passwort verschluesselt speichern
    $encryptedValue = Crypto::encrypt($value);
    
    // Einfuegen
    $id = Database::insert(
        'INSERT INTO known_passwords (password_type, password_value, description, created_by) VALUES (?, ?, ?, ?)',
        [$type, $encryptedValue, $description ?: null, $adminPayload['user_id']]
    );
    
    // Activity Log
    ActivityLogger::logAdmin($adminPayload, 'password_created', 'known_password', $id,
        "Neues {$type}-Passwort angelegt",
        ['password_type' => $type, 'description' => $description]
    );
    
    $password = Database::queryOne('SELECT * FROM known_passwords WHERE id = ?', [$id]);
    json_success(['password' => $password], 'Passwort angelegt');
}

/**
 * PUT /admin/passwords/{id}
 * Body: { password_value?: "...", description?: "...", is_active?: 0|1 }
 */
function handleUpdatePassword(int $id, array $adminPayload): void {
    $password = Database::queryOne('SELECT * FROM known_passwords WHERE id = ?', [$id]);
    if (!$password) {
        json_error('Passwort nicht gefunden', 404);
    }
    
    $data = get_json_body();
    $changes = [];
    
    // Passwort-Wert aendern
    if (isset($data['password_value'])) {
        $value = trim($data['password_value']);
        if (strlen($value) === 0) {
            json_error('Passwort darf nicht leer sein', 400);
        }
        // SV-006: Duplikat-Pruefung gegen entschluesselte Werte
        $allOfType = Database::query(
            'SELECT id, password_value FROM known_passwords WHERE password_type = ? AND id != ?',
            [$password['password_type'], $id]
        );
        foreach ($allOfType as $row) {
            $storedValue = $row['password_value'];
            try { $storedValue = Crypto::decrypt($storedValue); } catch (Exception $e) {}
            if ($storedValue === $value) {
                json_error('Dieses Passwort existiert bereits fuer diesen Typ', 409);
            }
        }
        // SV-006 Fix: Verschluesselt speichern
        $encryptedValue = Crypto::encrypt($value);
        Database::execute('UPDATE known_passwords SET password_value = ? WHERE id = ?', [$encryptedValue, $id]);
        $changes['password_value'] = '(geaendert)';
    }
    
    // Beschreibung aendern
    if (isset($data['description'])) {
        $desc = trim($data['description']);
        Database::execute('UPDATE known_passwords SET description = ? WHERE id = ?', [$desc ?: null, $id]);
        $changes['description'] = $desc;
    }
    
    // Aktiv-Status aendern
    if (isset($data['is_active'])) {
        $active = $data['is_active'] ? 1 : 0;
        Database::execute('UPDATE known_passwords SET is_active = ? WHERE id = ?', [$active, $id]);
        $changes['is_active'] = $active;
    }
    
    // Activity Log
    ActivityLogger::logAdmin($adminPayload, 'password_updated', 'known_password', $id,
        "Passwort bearbeitet (Typ: {$password['password_type']})",
        ['changes' => $changes]
    );
    
    $updated = Database::queryOne('SELECT * FROM known_passwords WHERE id = ?', [$id]);
    json_success(['password' => $updated], 'Passwort aktualisiert');
}

/**
 * DELETE /admin/passwords/{id}
 * Soft-Delete (is_active = 0).
 */
function handleDeletePassword(int $id, array $adminPayload): void {
    $password = Database::queryOne('SELECT * FROM known_passwords WHERE id = ?', [$id]);
    if (!$password) {
        json_error('Passwort nicht gefunden', 404);
    }
    
    if (!$password['is_active']) {
        json_error('Passwort ist bereits deaktiviert', 400);
    }
    
    Database::execute('UPDATE known_passwords SET is_active = 0 WHERE id = ?', [$id]);
    
    // Activity Log
    ActivityLogger::logAdmin($adminPayload, 'password_deleted', 'known_password', $id,
        "Passwort deaktiviert (Typ: {$password['password_type']})",
        ['password_type' => $password['password_type']]
    );
    
    json_success([], 'Passwort deaktiviert');
}
